### ISSUE TITLE
Captive Portal Check

### ISSUE THREAD

**Issue by IzzySoft** [mvglasow/satstat#171]
I got reports from users who complained to be tracked by SatStat, as it seemed to call out to Google servers (clientX.google.com). I initially thought that might have been the device itself looking for AGPS data from some supple*.google.com server. To rule out SatStat itself might be responsible, I searched the code – and found [this](https://github.com/mvglasow/satstat/blob/afd1e965ab42390dbd3ab6db74e23d475c941491/src/com/vonglasow/michael/satstat/utils/WifiCapabilities.java#L63):

    public static final String CAPTIVE_PORTAL_SERVER = "clients3.google.com";

An unfortunate choice for those who not want anything to do with "Google tracking" and because of that explicitly reached out to F-Droid. Would it be possible to substitute some "neutral server" – or disable captive portal check at all? As a maintainer at F-Droid I've been asked whether we should add an AntiFeature (NonFreeNet? Tracking?) – I'd rather not, and prefer having it solved in a better way :wink:

Thanks in advance!

**Comment by mvglasow** [comment]
Indeed, SatStat makes these calls to determine if it is behind a captive portal.

The underlying issue is that Android does not report any status on AGPS updates, so the best SatStat can do is to verify it actually has a working Internet connection that is not intercepted.

The method employed here is the same used by Android itself (even without GApps installed)—if one were to monitor the traffic generated by an Android device, the same servers would get polled whenever the device connects to a wifi network, even without SatStat installed. This is the reason I chose this route—the only footprint it generates is no different from what the device repeatedly generates anyway.

On Google’s end, the only data that should be revealed is the device’s public IP (shared with an entire network if it is behind a NAT gateway), the URL called and possibly a user agent string, though I would have to check what SatStat uses here. This is about as minimal as it gets, without sacrificing captive portal detection. Also, I (or SatStat) have no way to get at this data, nor do I have any intention.

Other vendors (Apple, Microsoft) also operate servers offering captive portal detection—but if SatStat were to use one of those, that would make it even more conspicuous than it is now (same device asking two different services), at least for someone who is monitoring the device’s Internet connection. (Things may look different for someone monitoring one of the servers.)

Of course, it is always possible to disable AGPS data updates, which will also stop captive portal detection by SatStat, and thus requests to these servers.

How would you suggest solving this (bearing in mind that reliable captive portal detection is a prerequisite for AGPS data download and cannot simply be disabled without breaking that)? As long as everything continues to work as it should, I am open to solutions.

**Comment by IzzySoft** [comment]
Thanks a lot for your explanations!

The idea seems sound to me (I've already given the "reporter" the link to this issue, so he can follow up) – and based on your explanations, I see no reason for an AntiFeature. Maybe until a better way is found, this should at least be mentioned in the app description? Like a third bullet point in the "limitations" list:

> * to update AGPS data, SatStat needs to check if it has a working Internet connection that is not intercepted. To keep this as unsuspicious as possible, SatStat polls the same server Android itself uses – but which belongs to Google. If you don't want SatStat calling out to a Google server for that check, please disable AGPS data updates for now. We're open to alternative solutions as long as everything continues to work as it should.

Unfortunately, I'm not aware of any alternative CP-URL working for everybody without the issue described. But after all: there's the server the AGPS data is retrieved from. Maybe that has a small textfile SatStat could check upon (retrieve it, check size/content – if OK, there obviously must be a working internet connection)?

I never dug into that, so I must admit I don't even know whether SatStat downloads the AGPS data itself or simply triggers Android's mechanism to do so. In the latter case, it's probably irrelevant SatStat makes that CP check if that is followed by another call to the same server in order to obtain the data (and that bullet point should be phrased slightly different).

**Comment by sm4rk0** [comment]
Maybe [this](https://stackoverflow.com/questions/37092500/is-the-string-captive-portal-detected-from-networkinfo-getextrainfo-a-reliab) can help? It makes sense that Android tracks the captive portal status.

**Comment by mvglasow** [comment]
@sm4rk0 the gist is that the property mentioned here is at the whim of the lower network layers, and not guaranteed to be available on any device. The other solution linked there is what SatStat currently does.

**Comment by mvglasow** [comment]
@IzzySoft SatStat triggers an Android mechanism to download AGPS data. The server used may vary between devices, and there is no return status or any other way to track if the operation was successful. Therefore, our best bet is to first determine if we have a working Internet connection, then fire off the command and just assume it works. Still far from perfect but the best we can get.

**Comment by gdt** [comment]
I am still not really following this.  I think there are a bunch of separate issues, and it might be good to separate them more clearly.

* The AGPS reload mechanism itself is at least a minor privacy issue and potentially much worse (if it includes a location).  This isn't satstat's fault but if there is a "trigger it" option it would be good to have this be clear.

* The "reload AGPS" menu item is what we're presumably talking about.   One way to deal with this might be to have it skip the check and always result in a toast "Request made; API does not provide any indication of success." and leave  it at that.  After all, normal people do not use satstat in the first place, and they really don't use this item.   And, people's phones are usually on the network, and rarely stuck in captive portals, because that interferes with other things.   If that's not ok, then a setting "perform captive portal check on agps reload request" could exist, and I'd argue it should default to off. 

* There is a setting about AGPS reload frequency.  But I don't understand if satstat starts on boot to do this. There is no "never" option (and it's not the default).   There is a "last update" displayed, and surely that's "last time satstat requested an update and we have no idea when the last actual update was".   (I am able to tell satstat to not request updates by unchecking all the networks, but that's non-obvious.)

I would expect that the underlying OS does some sort of AGPS fetching.   (Presumably the AGPS data is ephemeris, to enable a fix on acquiring a signal without waiting for 30s.)    So I don't really understand why there is any need for periodic initiation of this by satstat.

On the main github page, I see screenshots but not a clear explanation of what the program does.  Please note that I am not trying to imply that anyone is being even the slightest bit devious.   There is a general problem in the android world (mobile world in general) that starts with "apps should be such that you don't need documentation", which is fine, but which omits "apps should document exactly what they do in terms of data storage and communication".   In this case, I suspect 99% of the real issues are in the base OS, not this app, but it would be nice to share any clarity about that, partly because setting an example is good for the world.

**Comment by IzzySoft** [comment]
Maybe good ideas would be:

* introducing an option "use captive portal check" (default: off) with a short description of its background, including that this would call a Google server (which in most cases would happen anyway on ~80% of the devices when Android pulls the data from suppl.google.com – with the exception of the other ~20% where manufacturers have configured their own SUPPL server in `gps.conf`)
* adjusting the AGPS reload frequency setting to include "never"
* whenever the reload is triggered manually (and captive portal check is off), show a toast as @gdt suggested
* nice-to-have in this context: an option to suppress that toast for those being annoyed of it :wink:

That, together with the update of the app's description I suggested above (maybe also in the Readme) should cover the issue.

**Comment by barbeau** [comment]
FWIW, in GPSTest I show a success/failure message based solely off the boolean returned from the method used to send these commands to the `LocationManager`:
https://developer.android.com/reference/android/location/LocationManager#sendExtraCommand(java.lang.String,%20java.lang.String,%20android.os.Bundle)

The description of the boolean returned is `true if the command succeeds.`

There are so many other variables is terms of whether these commands actually impact the state of the handset that IMHO doing a captive portal check wasn't worth it.

Also, injecting time and XTRA and clearing assist data is done only on-demand in GPSTest when the user taps the button in the nav drawer.
